### PR TITLE
TYP: enable disallow_untyped_decorators

### LINKED ISSUE #33455: TYP: enable disallow-untyped-decorators in CI
xref https://github.com/pandas-dev/pandas/pull/33160#issuecomment-608707397

in order to ensure adding decorators or changing existing decorators don't lose types on our typed functions, we could enable this flag, after resolving the current violations. (expand details)

<details>

$ mypy pandas --disallow-untyped-decorators
pandas\core\window\numba_.py:58: error: Untyped decorator makes function "roll_apply" untyped
pandas\tseries\offsets.py:458: error: Untyped decorator makes function "freqstr" untyped
pandas\core\nanops.py:97: error: Untyped decorator makes function "f" untyped
pandas\core\nanops.py:477: error: Untyped decorator makes function "nansum" untyped
pandas\core\nanops.py:799: error: Untyped decorator makes function "nansem" untyped
pandas\core\nanops.py:849: error: Untyped decorator makes function "reduction" untyped
pandas\core\nanops.py:880: error: Untyped decorator makes function "nanargmax" untyped
pandas\core\nanops.py:916: error: Untyped decorator makes function "nanargmin" untyped
pandas\core\nanops.py:952: error: Untyped decorator makes function "nanskew" untyped
pandas\core\nanops.py:1036: error: Untyped decorator makes function "nankurt" untyped
pandas\core\nanops.py:1129: error: Untyped decorator makes function "nanprod" untyped
pandas\core\nanops.py:1309: error: Untyped decorator makes function "nancorr" untyped
pandas\core\nanops.py:1364: error: Untyped decorator makes function "nancov" untyped
pandas\core\apply.py:136: error: Untyped decorator makes function "dtypes" untyped
pandas\tseries\frequencies.py:333: error: Untyped decorator makes function "is_unique" untyped
pandas\tseries\frequencies.py:337: error: Untyped decorator makes function "is_unique_asi8" untyped
pandas\core\arrays\period.py:270: error: Untyped decorator makes function "dtype" untyped
pandas\core\base.py:198: error: Untyped decorator makes function "ndim" untyped
pandas\core\indexes\base.py:1619: error: Untyped decorator makes function "is_unique" untyped
pandas\core\indexes\base.py:1973: error: Untyped decorator makes function "inferred_type" untyped
pandas\core\indexes\base.py:1980: error: Untyped decorator makes function "is_all_dates" untyped
pandas\core\indexes\base.py:2021: error: Untyped decorator makes function "hasnans" untyped
pandas\core\indexes\base.py:3840: error: Untyped decorator makes function "array" untyped
pandas\core\arrays\integer.py:62: error: Untyped decorator makes function "is_signed_integer" untyped
pandas\core\arrays\integer.py:66: error: Untyped decorator makes function "is_unsigned_integer" untyped
pandas\core\arrays\integer.py:74: error: Untyped decorator makes function "numpy_dtype" untyped
pandas\core\arrays\integer.py:79: error: Untyped decorator makes function "kind" untyped
pandas\core\arrays\integer.py:83: error: Untyped decorator makes function "itemsize" untyped
pandas\core\arrays\integer.py:341: error: Untyped decorator makes function "dtype" untyped
pandas\core\arrays\categorical.py:497: error: Untyped decorator makes function "size" untyped
pandas\core\arrays\categorical.py:504: error: Untyped decorator makes function "itemsize" untyped
pandas\core\indexes\numeric.py:452: error: Untyped decorator makes function "is_unique" untyped
pandas\core\indexes\range.py:170: error: Untyped decorator makes function "_int64index" untyped
pandas\core\indexes\range.py:280: error: Untyped decorator makes function "nbytes" untyped
pandas\core\indexes\range.py:325: error: Untyped decorator makes function "is_monotonic_increasing" untyped
pandas\core\indexes\range.py:329: error: Untyped decorator makes function "is_monotonic_decreasing" untyped
pandas\core\indexes\multi.py:1133: error: Untyped decorator makes function "dtype" untyped
pandas\core\indexes\multi.py:1152: error: Untyped decorator makes function "nbytes" untyped
pandas\core\indexes\multi.py:1414: error: Untyped decorator makes function "inferred_type" untyped
pandas\core\indexes\multi.py:1450: error: Untyped decorator makes function "is_monotonic_increasing" untyped
pandas\core\indexes\multi.py:1474: error: Untyped decorator makes function "is_monotonic_decreasing" untyped
pandas\core\indexes\datetimes.py:279: error: Untyped decorator makes function "_is_dates_only" untyped
pandas\core\indexes\period.py:348: error: Untyped decorator makes function "_int64index" untyped
pandas\core\indexes\interval.py:384: error: Untyped decorator makes function "_multiindex" untyped
pandas\core\indexes\interval.py:388: error: Untyped decorator makes function "values" untyped
pandas\core\indexes\interval.py:430: error: Untyped decorator makes function "is_monotonic_decreasing" untyped
pandas\core\indexes\interval.py:1031: error: Untyped decorator makes function "intersection" untyped
pandas\core\groupby\ops.py:257: error: Untyped decorator makes function "is_monotonic" untyped
pandas\core\groupby\ops.py:270: error: Untyped decorator makes function "codes_info" untyped
pandas\core\groupby\ops.py:288: error: Untyped decorator makes function "ngroups" untyped
pandas\core\groupby\ops.py:298: error: Untyped decorator makes function "result_index" untyped
pandas\core\groupby\ops.py:818: error: Untyped decorator makes function "reconstructed_codes" untyped
pandas\core\groupby\grouper.py:395: error: Untyped decorator makes function "result_index" untyped
pandas\core\groupby\grouper.py:420: error: Untyped decorator makes function "groups" untyped
pandas\core\window\rolling.py:1802: error: Untyped decorator makes function "is_datetimelike" untyped
pandas\core\window\rolling.py:1808: error: Untyped decorator makes function "_on" untyped
pandas\tests\frame\methods\test_replace.py:14: error: Untyped decorator makes function "mix_ab" untyped
pandas\tests\frame\methods\test_replace.py:19: error: Untyped decorator makes function "mix_abc" untyped
Found 58 errors in 20 files (checked 987 source files)

</details>

### LINKED ISSUE #41588: TYP: IntervalIndex.right/left might be infered as function
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```py
import pandas as pd

index: pd.IntervalIndex = pd.IntervalIndex.from_arrays([1], [2])
reveal_type(index)
reveal_type(index.right)
index.right.array
```

pyright
```
info: Type of "index" is "IntervalIndex"
info: Type of "index.right" is "() -> Index"
error: Cannot access member "array" for type "function"
    Member "array" is unknown (reportFunctionMemberAccess)
```

mypy
```
note: Revealed type is 'pandas.core.indexes.interval.IntervalIndex'
note: Revealed type is 'Any'
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.9.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.13.0-36-generic
Version          : #40~16.04.1-Ubuntu SMP Fri Feb 16 23:25:58 UTC 2018
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.4
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.23.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 2021.05.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### LINKED ISSUE #43387: TYP:  RangeIndex._constructor
`RangeIndex._constructor` returns `Type[Int64Index]` but it is supposed to return the current type of the class itself (`Type[RangeIndex]`) .

pandas/core/indexes/range.py:179: error: Return type "Type[Int64Index]" of "_constructor" incompatible with return type "Type[RangeIndex]" in supertype "Index"  [override]

Found by declaring `cache_readonly = property`.

### LINKED ISSUE #43388: TYP:  IndexEngine.get_indexer_with_fill
pandas/core/indexes/base.py:3696: error: "IndexEngine" has no attribute "get_indexer_with_fill"  [attr-defined]

Is this code intended to be reached only by certain sub-classes of `IndexEngine` that have `get_indexer_with_fill`?

I'm not sure how this is related to `cache_readonly = property` but it was found that way.

### PR DESCRIPTION
**Pull request by twoertwein**
- [ ] closes #33455, closes #41588, (closes #43388,  closes #43387)
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

Some of the many type errors unveiled by `cache_readonly` might be easily fixable. This PR simply ignores all the uncovered errors to quickly enable mypy's `disallow_untyped_decorators` to prevent future untyped decorators.

I enabled mypy's `disallow_untyped_decorators` but pyright's `reportUntypedFunctionDecorator` finds the following three functions that are not found by mypy, so I didn't enable `reportUntypedFunctionDecorator` (can't add ignore statements, otherwise mypy will complain about unused ignore statements, edit: using pyright's file-based ignore comments):

- pandas/conftest.py:1505
- ~~pandas/core/util/numba_.py:93~~
- pandas/core/window/numba_.py:244

### CODE REVIEW

File: pandas/_libs/properties.pyx [hunk 1]
```diff
@@ -12,10 +12,10 @@ cdef class CachedProperty:
     cdef readonly:
         object func, name, __doc__
 
-    def __init__(self, func):
-        self.func = func
-        self.name = func.__name__
-        self.__doc__ = getattr(func, '__doc__', None)
+    def __init__(self, fget):
```
Comment (twoertwein) [inline pandas/_libs/properties.pyx:15]: `func` -> `fget` to be more `property`-like

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -3623,7 +3623,12 @@ def _get_indexer(
         else:
             tgt_values = target._get_engine_target()
             if target._is_multi and self._is_multi:
-                tgt_values = self._engine._extract_level_codes(target)
+                # error: "IndexEngine" has no attribute "_extract_level_codes"
+                tgt_values = (
+                    self._engine._extract_level_codes(  # type: ignore[attr-defined]
```
Comment (twoertwein) [inline pandas/core/indexes/base.py:3628]: How do I make this black&mypy&flake8 compatible? Black's formatting moves it where it isn't helpful for mypy.  Having `tgt_values = self._engine._extract_level_codes(  # type: ignore[attr-defined]` in one line is too long for flake8.

File: pandas/_libs/properties.pyi [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    Any,
+    Type,
+)
+
+cache_readonly = property
+
+class AxisProperty:
+    def __init__(self, axis: int = ..., doc: str = ...) -> None: ...
+    def __get__(self, obj: Any, typ: Type) -> Any: ...  # TODO
+    def __set__(self, obj: Any, value: Any) -> None: ...  # TODO
```
Comment (twoertwein) [inline pandas/_libs/properties.pyi:11]: `AxisProperty` has to be in this file as pyi files are assumed to be complete.

**Comment by twoertwein** [comment]
> yah there are several methods near the top of DatetimeIndex that are there instead of using the decorator for this reason, most of the "# methods that dispatch to DatetimeArray and wrap result" section

So the solution is to write a doc-string?

I'm surprised that this warning/error occurs only for this PR. The 1.3.3 [documentation of DatetimeIndex](https://pandas.pydata.org/docs/reference/api/pandas.DatetimeIndex.html) has the same issue: 
image

**Comment by twoertwein** [comment]
The decorator seems to handle doc-strings correctly, all other attributes (for example `tz`)  have their doc-string string in the web-documentation. The main difference between `tz` and `std` is that `DatetimeArray.tz` has a doc-string while `DatetimeArray.std` has none: `DatetimeIndex.tz` has therefore a doc-string but `DatetimeArray.std` doesn't have a doc-string.

I will write a doc-string in a separate PR.

**Comment by twoertwein** [comment]
Should be ready for review :)

**Review by simonjayhawkins — APPROVED** [review]
Thanks @twoertwein very nice. some comments and suggestions but none blockers.

**Comment by twoertwein** [comment]
Rebased and addressed the big comments.

If it is green, it should be ready for merging from my perspective.

**Comment by twoertwein** [comment]
I think numba analyzes the source code to JIT it. I think the errors are a results of jitting a function that itself has a jitted function where the inner function is jitted with `typed_numba_jit`.

**Comment by simonjayhawkins** [comment]
@twoertwein can you resolve conflicts

**Comment by simonjayhawkins** [comment]
> I removed the numba wrapper and rebased

i've opened #44233 that we could maybe merge after this.

**Comment by twoertwein** [comment]
> i've opened #44233 that we could maybe merge after this.

Great! I'll rebase this PR to resolve the conflicts with master.

File: pandas/_libs/properties.pyx (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/_libs/properties.pyx:15]: So this helps catch typing errors of `cache_readonly` since `cache_readonly = property`? I'll mark as resolved.

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -2699,10 +2699,9 @@ def _get_codes_for_values(values, categories: Index) -> np.ndarray:
     """
     dtype_equal = is_dtype_equal(values.dtype, categories.dtype)
 
-    if is_extension_array_dtype(categories.dtype) and is_object_dtype(values):
+    if isinstance(categories.dtype, ExtensionDtype) and is_object_dtype(values):
         # Support inferring the correct extension dtype from an array of
         # scalar objects. e.g.
-        # Categorical(array[Period, Period], categories=PeriodIndex(...))
```
Comment (simonjayhawkins) [inline pandas/core/arrays/categorical.py:2705]: was this removal intentional?

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -4677,7 +4682,8 @@ def values(self) -> ArrayLike:
         """
         return self._data
 
-    @cache_readonly
+    # error: Decorated property not supported
+    @cache_readonly  # type: ignore[misc]
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py:4686]: IIRC there is a mypy issue for this. can you add a link to that issue here as a comment.

Comment (twoertwein) [reply pandas/core/indexes/base.py:4687]: linked the issue

File: pandas/core/arrays/categorical.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/arrays/categorical.py:2705]: yes

Comment (simonjayhawkins) [reply pandas/core/arrays/categorical.py:2705]: probably should remove the `e.g.` on the previous line as well?

Comment (twoertwein) [reply pandas/core/arrays/categorical.py:2705]: Sorry, I meant it was not intentional

Comment (simonjayhawkins) [reply pandas/core/arrays/categorical.py:2705]: I see it has been restored in the latest commit. marking as resolved.

Comment (twoertwein) [reply pandas/core/arrays/categorical.py:2705]: Finally green :)

**Review by simonjayhawkins — APPROVED** [review]

**Comment by jreback** [comment]
nice thanks @twoertwein and @simonjayhawkins for the review.
